File: voiceprint/http_server/speakers.py
from flask import Blueprint, jsonify, request, current_app

speakers = Blueprint('speakers', __name__)

@speakers.route('', methods=['GET'])
def get_speakers():
  speakerService = current_app.config.get('SPEAKER_SERVICE')
  speakers = speakerService.get_speakers()
  return jsonify(speakers), 200

@speakers.route('', methods=['POST'])
def add_speaker():
  if not request.json or 'name' not in request.json:
    return jsonify({"error": "Missing speaker name"}), 400

  speaker_name = request.json['name']

  speakerService = current_app.config.get('SPEAKER_SERVICE')
  speakerService.add_speaker(speaker_name)

  return jsonify({"message": "Speaker added successfully"}), 201

@speakers.route('/<speaker_name>/samples', methods=['POST'])
def add_speaker_sample(speaker_name):
  if not request.files or 'file' not in request.files:
    return jsonify({"error": "No file part"}), 400

  file = request.files['file']
  if file.filename == '':
    return jsonify({"error": "No selected file"}), 400

  speakerService = current_app.config.get('SPEAKER_SERVICE')
  speakerService.add_speaker_sample(speaker_name, file)

  return jsonify({"message": "Speaker sample added successfully"}), 201

@speakers.route('/<speaker_name>/samples/<sample_name>', methods=['DELETE'])
def delete_speaker_sample(speaker_name, sample_name):
  speakerService = current_app.config.get('SPEAKER_SERVICE')
  speakerService.delete_speaker_sample(speaker_name, sample_name)
  return jsonify({"message": "Speaker sample deleted successfully"}), 200

File: frontend/api/speakers.ts
"use server";

import ky from "ky";
import { API_BASE_URL } from "./constants";
import { revalidatePath } from "next/cache";

export type SpeakersData = Record<string, string[]>;

export async function getSpeakers() {
  const res = await ky.get<SpeakersData>(`${API_BASE_URL}/speakers`);
  return res.json();
}

export async function addSpeaker({ name }: { name: string }) {
  "use server";
  await ky.post(`${API_BASE_URL}/speakers`, { json: { name } }).json();

  revalidatePath("/speakers");
}

export async function addSpeakerSample({
  speakerName,
  file,
}: {
  speakerName: string;
  file: File;
}) {
  "use server";
  const formData = new FormData();
  formData.append("file", file);
  await ky
    .post(`${API_BASE_URL}/speakers/${speakerName}/samples`, { body: formData })
    .json();

  revalidatePath("/speakers");
}

export async function deleteSpeakerSample({
  speakerName,
  sampleName,
}: {
  speakerName: string;
  sampleName: string;
}) {
  "use server";
  await ky.delete(
    `${API_BASE_URL}/speakers/${speakerName}/samples/${sampleName}`
  );
  revalidatePath("/speakers");
}
